File: backend/app/schemas/excel.py
"""Pydantic schemas for Excel file upload and parsing."""
from datetime import date
from pydantic import BaseModel, ConfigDict, Field


class ExcelRowData(BaseModel):
    """Schema representing a single row from the Excel file."""
    model_config = ConfigDict(populate_by_name=True)

    project: str = Field(..., alias="Project")
    deliverable: str = Field(..., alias="Deliverable")
    due_date: date = Field(..., alias="Due Date")
    frequency: str = Field(..., alias="Frequency")
    project_manager: str = Field(..., alias="Project Manager")


class ExcelValidationError(BaseModel):
    """Schema for validation errors during Excel parsing."""
    row: int
    column: str
    value: str | None
    error: str


class ExcelUploadResult(BaseModel):
    """Schema for Excel upload results."""
    success: bool
    filename: str
    total_rows: int
    imported_rows: int
    skipped_rows: int
    errors: list[ExcelValidationError] = []
    projects_created: int = 0
    managers_created: int = 0
    deliverables_created: int = 0


class ExcelPreviewRow(BaseModel):
    """Schema for previewing Excel data before import."""
    row_number: int
    project: str
    deliverable: str
    due_date: str
    frequency: str
    project_manager: str
    is_valid: bool
    validation_errors: list[str] = []


class ExcelPreviewResult(BaseModel):
    """Schema for Excel preview results."""
    filename: str
    total_rows: int
    valid_rows: int
    invalid_rows: int
    preview_data: list[ExcelPreviewRow]
    column_mapping: dict[str, str]

File: frontend/src/types/excel.ts
export interface ExcelValidationError {
  row: number
  column: string
  value: string | null
  error: string
}

export interface ExcelUploadResult {
  success: boolean
  filename: string
  total_rows: number
  imported_rows: number
  skipped_rows: number
  errors: ExcelValidationError[]
  projects_created: number
  managers_created: number
  deliverables_created: number
}

export interface ExcelPreviewRow {
  row_number: number
  project: string
  deliverable: string
  due_date: string
  frequency: string
  project_manager: string
  is_valid: boolean
  validation_errors: string[]
}

export interface ExcelPreviewResult {
  filename: string
  total_rows: number
  valid_rows: number
  invalid_rows: number
  preview_data: ExcelPreviewRow[]
  column_mapping: Record<string, string>
}

export interface ExcelTemplateResponse {
  columns: Record<string, { description: string; type: string; required?: boolean; valid_values?: Record<string, string>; default?: string; example?: string; format?: string }>
  notes: string[]
}
